### ISSUE TITLE
Add Check Support for Java 17 Sealed Classes: New Check UnnecessaryPermitsClause

### ISSUE THREAD

**Issue by mahfouz72** [checkstyle/checkstyle#14986]
child of #14969 

sealed classes are useful if there is a need to define a strict hierarchy of extending classes. To mention all the allowed subclasses, there is a keyword permits, which subclasses' names should follow.
In case when all subclasses are declared in the same file there is no need to mention them explicitly and `permits` clause can be omitted.

---

**Config**
```XML
$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="UnnecessaryPermitsClause"/>
    </module>
</module>
```

---

**Example1**
```Java
// violation below, 'no need for permitted types because all subclasses are in the same file'
sealed class A permits B, C, D, E {} 
final class B extends A {}
final class C extends A {}
final class D extends A {}
final class E extends A {}
```
---

**Example2**
```Java
sealed class A permits B, C, D, E {} // ok, subclass E is in another file so permits is necessary
final class B extends A {}
final class C extends A {}
final class D extends A {}

```
---
**Expected Output**
```
$ java -jar checkstyle-10.19.0-all.jar -c config.xml SealedClasses.java 
Starting audit...
[ERROR] D:\SealedClasses.java:1:1: no need for permitted types all subclasses are in the same file
Audit done.
Checkstyle ends with 1 errors.
```

**Comment by rnveach** [comment]
> no need for permitted types all subclasses are in the same file

I suggest something like "Unnecessary 'permits' and class list when all subclasses are in the same file."

**Comment by nrmancuso** [comment]
How about `Unnecessary 'permits' clause for classes in the same compilation unit`?

**Comment by nrmancuso** [comment]
I am going to go ahead and approve this issue, we can discuss the message further in here if need be.

**Comment by Zopsss** [comment]
Due to the limitations of CS: https://checkstyle.org/writingchecks.html#Limitations

> You cannot see the contents of other files. You have access to the content of one file only during all Checks execution. All files are processed one by one.


Is it possible to implement this check?

Suppose if the permitted class is in other file, how can we check if that class is part of the permitted classes or not?

**Comment by Atharv3221** [comment]
> Due to the limitations of CS: https://checkstyle.org/writingchecks.html#Limitations
> 
> > You cannot see the contents of other files. You have access to the content of one file only during all Checks execution. All files are processed one by one.
> 
> 
> Is it possible to implement this check?
> 
> Suppose if the permitted class is in other file, how can we check if that class is part of the permitted classes or not?
> 

We don't exactly need to go in other files, I thought implementing this check by

1) Count number of classes, interfaces that permitted 

2) If we get classes which extends that class or an interface which implements then we subtract count by 1

3) As a result if all are in same file the count will eventually be 0 and if not then it won't be zero 

4) if count 0 we can give violation to that permit clause can be removed 

All classes which are permitted do need to extend the sealed class or code won't even compile.

**Comment by nrmancuso** [comment]
> Suppose if the permitted class is in other file, how can we check if that class is part of the permitted classes or not?

@Zopsss, please explain why this is required.

**Comment by Zopsss** [comment]
As shown in example 2:

```java
sealed class A permits B, C, D, E {} // ok, subclass E is in another file so permits is necessary
final class B extends A {}
final class C extends A {}
final class D extends A {}
```

Class E is in another file and it is defined in the permits keyword so it can extend class A, no violation for it. So suppose, if there's another class F which is not defined in permits keyword and extends sealed class A then we need to log it's violation. And due to the limitations of CS, I doubt this is possible.

**Comment by Atharv3221** [comment]
> As shown in example 2:
> 
> ```java
> sealed class A permits B, C, D, E {} // ok, subclass E is in another file so permits is necessary
> final class B extends A {}
> final class C extends A {}
> final class D extends A {}
> ```
> 
> Class E is in another file and it is defined in the permits keyword so it can extend class A, no violation for it. So suppose, if there's another class F which is not defined in permits keyword and extends sealed class A then we need to log it's violation. And due to the limitations of CS, I doubt this is possible. 

But it's not possible @Zopsss 
If for sealed class that class isn't permitted it can't extend it in another file

**Comment by Zopsss** [comment]
> If for sealed class that class isn't permitted it can't extend it in another file


Ah got it! My bad I misunderstood the issue description

**Comment by Atharv3221** [comment]
My approach is correct or need to think of another solution ?

**Comment by nrmancuso** [comment]
> My approach is correct or need to think of another solution ?

Your idea is solid, however I would consider using the names of the inner classes as opposed to an integer, and we may need to track what class is extended.

**Comment by Atharv3221** [comment]
> > My approach is correct or need to think of another solution ?
> 
> Your idea is solid, however I would consider using the names of the inner classes as opposed to an integer, and we may need to track what class is extended, so we might want to attach some metadata to the class name.

I will create a class with class name and DetailAST will use it to for instances for sealed class and map it with the count, any more metadata can also be added in that instance.

**Comment by MonuChaudhary14** [comment]
Can i work on this issue?

**Comment by Atharv3221** [comment]
I wish to continue work on this but my exams are going on

**Comment by Atharv3221** [comment]
@romani 
I am not getting we already have a check https://checkstyle.sourceforge.o/checks/design/sealedshouldhavepermitslist.html#SealedShouldHavePermitsList
according to which all sealed classes should have permit list but according to this check idea we shouldn't if all permitted classes are in same file, it would create problem when used together.

**Comment by mahfouz72** [comment]
Yes, we have multiple cases where checks can conflict, and the user should decide which one they want to follow.

**Comment by romani** [comment]
@M-SaaD-H, if you have time please look at https://github.com/checkstyle/checkstyle/pull/18702 it might be easy win to finish it

**Comment by M-SaaD-H** [comment]
on it.

**Comment by romani** [comment]
Fix is merged
